File: pset6/credit/credit.py
from cs50 import get_int
import sys

# get positive integer
number = 0
while number <= 0:
    number = get_int("Number: ")

s_number = str(number)
length = len(s_number)

# check for valid length
lengths = [13, 15, 16]
if length not in lengths:
    print("INVALID")
    sys.exit(1)

# Luhn's algorithm
total = 0

# add every other digit, starting with number's last digit
for c in range(length - 1, -1, -2):
    total += int(s_number[c])

# multiply every other digit by 2, starting with the number’s second-to-last digit
# then, add every digit of the results
for c in range(length - 2, -1, -2):
    x = 2 * int(s_number[c])
    for i in str(x):
        total += int(i)

# check for Luhn's algorithm correctness
if total % 10 != 0:
    print("INVALID")
    sys.exit(1)

# find out the company
mastercard = "51", "52", "53", "54", "55"
amex = "34", "37"
visa = "4"

if s_number[:2] in mastercard and length == 16:
    print("MASTERCARD")
elif s_number[:2] in amex and length == 15:
    print("AMEX")
elif s_number[:1] in visa and length in [13, 16]:
    print("VISA")
else:
    print("INVALID")
    sys.exit(1)

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    // prompt for imput
    long N = get_long("Number: ");

    // 13-digit number
    if (N >= pow(10, 12) && N < pow(10, 13))
    {
        // calculate checksum
        int sum = 0;

        for (int i = 1; i < 12; i += 2)
        {
            int x = 2 * (N / (long)pow(10, i) % 10);
            if (x < 10)
            {
                sum += x;
            }
            else
            {
                sum += x / 10;
                sum += x % 10;
            }
        }

        for (int i = 0; i < 13; i += 2)
        {
            sum += N / (long)(pow(10, i)) % 10;
        }

        if (sum % 10 == 0 && N / (long)pow(10, 12) == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    // 15-digit number
    else if (N >= (long)pow(10,  14) && N < (long)pow(10, 15))
    {
        // calculate checksum
        int sum = 0;

        for (int i = 1; i < 14; i += 2)
        {
            int x = 2 * (N / (long)pow(10, i) % 10);
            if (x < 10)
            {
                sum += x;
            }
            else
            {
                sum += x / 10;
                sum += x % 10;
            }
        }

        for (int i = 0; i < 15; i += 2)
        {
            sum += N / (long)pow(10, i) % 10;
        }

        int first_two = N / (long)pow(10, 13);

        if (sum % 10 == 0 && (first_two == 34 || first_two == 37))
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    // 16-digit number
    else if (N >= (long)pow(10, 15) && N < (long)pow(10, 16))
    {
        // calculate checksum
        int sum = 0;

        for (int i = 1; i < 16; i += 2)
        {
            int x = 2 * (N / (long)pow(10, i) % 10);
            if (x < 10)
            {
                sum += x;
            }
            else
            {
                sum += x / 10;
                sum += x % 10;
            }
        }

        for (int i = 0; i < 15; i += 2)
        {
            sum += N / (long)pow(10, i) % 10;
        }

        int first_two = N / pow(10, 14);

        if (sum % 10 == 0 && N / (long)pow(10, 15) == 4)
        {
            printf("VISA\n");
        }
        else if (sum % 10 == 0 && (first_two == 51 || first_two == 52 || first_two == 53 || first_two == 54 || first_two == 55))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}